store/entity-store.html

```diff
@@ -84,92 +123,107 @@
 		// updating the UI more consistently when dependent entities change as a result of Siren
 		// actions.
 		fetch: function(entityId, token, bypassCache) {
-			const lowerCaseToken = token.toLowerCase();
-			const lowerCaseEntityId = entityId.toLowerCase();
+			return this._getToken(token).then(function(resolved) {
+				const cacheKey = resolved.cacheKey;
+				const tokenValue = resolved.tokenValue;
 
-			var entity = this._initContainer(this._store, entityId, token);
-			if (!entity || bypassCache) {
+				const lowerCaseToken = cacheKey.toLowerCase();
+				const lowerCaseEntityId = entityId.toLowerCase();
 
-				var headers = new Headers();
-				token && headers.set('Authorization', 'Bearer ' + token);
+				var entity = this._initContainer(this._store, entityId, cacheKey);
+				if (!entity || bypassCache) {
 
-				if (bypassCache) {
-					headers.set('pragma', 'no-cache');
-					headers.set('cache-control', 'no-cache');
-				}
+					var headers = new Headers();
+					headers.set('Authorization', 'Bearer ' + tokenValue);
```

Comment:
I think the original `token && headers.set(...` was intended to allow users of the component to default to using a session based token instead of passing a token in. I think `d2l-fetch-auth` will only use the LMS token API to swap the user's cookie for a token if there is no existing `Authorization` header. So by passing in a `token=''` attribute you could utilize that default token mechanism.

We unfortunately made use of this capability in rubric editing to take advantage of the in-built d2l-fetch-auth token expiry handling to support long editing sessions.  Mostly because we have not had time to introduce something like the promise based expiry that you are introducing.

So I would be interested in moving back to a scenario where we pass in a token and manage the token expiry using the promise mechanism you are introducing if we can make that work with our usage scenario.

How are you managing the expiry of tokens? What is the usage context for the component you are including this feature for?

Reply:
For us, our we created a computed `_getToken` that looks like this:

```
_getToken(token) {
	return () => { return Promise.resolve(token); };
}
```

We bind that through all of our children. Then we setup a `setInterval` that calls the [frau-jwt](https://github.com/Brightspace/frau-jwt) to get a new token every X amount of time. We're playing with what amount of time that will need to be (maybe expires / 2 or something?). When it sets `this.token` the computed call gets activated with a new token. Since we're binding our children to a function (Promise), we don't rebind and thus we don't re-draw everything. With our new cacheKey based on the decoded token, we don't need to refetch if we're getting the same claims back. Every time we *do* need to fetch, we'll call our promise which will return the latest and greatest token.

There are a few caveats that we've found. `frau-jwt` caches the token until it expires which is troublesome because we want a new token *before* it expires. To get around that we have to call `jwt._resetCaches()`. We also need to decode the token to find out the expiry in the app.

Reply:
So is this something that is embedded in your app's/web component's code? Is this something that is deployed through BSI? I'm just wondering if this is something that could be included in library code so it is generally usable by other BSI web components.  In previous discussions I've been involved in there has been talk of creating some sort of 'Auth' wrapper web component that will manage this on behave of the wrapped components.

I'd be interested in seeing the code that is utilizing this so I can get a better understanding of the strategy you are using.

And just to confirm, we'll need this to include the check for token existence prior to setting the auth header or it will break the rubrics editing components.

Reply:
I'm trying to figure out how that check ever failed. Prior to my change, the first thing this method did was: `const lowerCaseToken = token.toLowerCase();`. If `token` was every null or undefined, it would blow up there.

Reply:
Empty string, potentially?

Reply:
Hmm, okay. I'll have a number of other changes I'll need to make in `_getToken` since currently if no token is present I don't carry on.

Reply:
yep empty string

Reply:
I've added the check back. When I pass an empty string, all my API calls fail with a 403 so I assume you are doing something different that will cause it to work. It would be sweet if you could grab this PR and try with your repo to make sure that use case is covered since I can't reproduce it here.